irohad/validation/impl/stateful_validator_impl.cpp

```diff
@@ -53,41 +73,54 @@ namespace iroha {
       // Filter only valid transactions
       auto filter = [&temporaryWsv, checking_transaction](auto &acc,
                                                           const auto &tx) {
-        auto answer = temporaryWsv.apply(tx, checking_transaction);
+        auto answer =
+            temporaryWsv.apply(*tx->makeOldModel(), checking_transaction);
         if (answer) {
           acc.push_back(tx);
         }
         return acc;
       };
 
-      auto &txs = proposal.transactions;
+      auto &txs = proposal->transactions();
       decltype(txs) valid = {};
 
-      model::Proposal validated_proposal(
-          std::accumulate(txs.begin(), txs.end(), valid, filter));
-      validated_proposal.height = proposal.height;
+      auto valid_txs = std::accumulate(txs.begin(), txs.end(), valid, filter);
+
+      // rework validation logic, so that this conversion is not needed
+      auto valid_proto_txs =
+          valid_txs
+          | boost::adaptors::transformed([](const auto &polymorphic_tx) {
+              return *static_cast<const shared_model::proto::Transaction *>(
```

Comment:
maybe just `return *polymorphic_tx.operator->();`?

Reply:
No, then we will allocate interface::Transaction which is abstract. And yet, ProposalBuilder below requires list of proto::Transactions